compiler/damlc/daml-stdlib-src/DA/Internal/Prelude.daml

```diff
@@ -528,8 +528,14 @@ traceId : Show b => b -> b
 traceId x = trace x x
 
 -- | `debug x` prints `x` for debugging purposes.
-debug : (Show b, Applicative m) => b -> m ()
-debug x = trace x (pure ())
+debug : (Show b, Action m) => b -> m ()
+debug x =
+  -- NOTE (MK): We introduce an explicit indirection via >>=
+  -- to make sure that evaluating something like
+  -- `debug x >> pure ()` does not print the debug statement.
+  -- Instead it will only be printed when you execute the monadic
+  -- action.
+  pure () >>= \_ -> pure (trace x ())
```

Comment:
Under what circumstances is this lazier? Only if `>>=` involved some sort of function composition, as in `State` or `Reader` and in `Update` or `Scenario` because they get rewritten into some sort of `Reader` in the speedy compiler, right? `Optional` and `Either _` are unaffected by this. Should we add some documentation around this? Or should we add an internal type class for monads that allow for `debug`?

Reply:
Given that I constantly get confused by strictness myself, I think trying to explain this in docs is probably going to do more harm than good. For `Either` it works as expected with and without this change but I think I managed to confuse myself sufficiently now to not be able to tell you why exactly :slightly_smiling_face: